examples/advanced_startup.py

```diff
@@ -0,0 +1,74 @@
+# This example covers advanced startup options and uses some real world examples for why you may need them.
+
+import asyncio
+import logging
+import os
+
+from typing import Optional
+
+import asyncpg  # asyncpg is not a dependency of the discord.py, and is only included here for illustrative purposes.
+import discord
+from discord.ext import commands
+from aiohttp import ClientSession
+
+
+class CustomBot(commands.Bot):
+    def __init__(
+        self,
+        *args,
+        initial_extensions: list,
+        db_pool: asyncpg.Pool,
+        web_client: ClientSession,
+        testing_guild_id: Optional[int] = None,
+        **kwargs,
+    ):
+        super().__init__(*args, **kwargs)
+        self.db_pool = db_pool
+        self.web_client = web_client
+        self.testing_guild_id = testing_guild_id
+        self.initial_extensions = initial_extensions
+
+    async def setup_hook(self) -> None:
+
+        # here, we are loading extensions prior to sync to ensure we are syncing interactions defined in those extensions.
+
+        for extension in self.initial_extensions:
+            await self.load_extension(extension)
+
+        # In overriding setup hook,
+        # we can do things that require a bot prior to starting to process events from the websocket.
+        # In this case, we are using this to ensure that once we are connected, we sync for the testing guild.
+        # You should not do this for every guild or for global sync, those should only be synced when changes happen.
+        if self.testing_guild_id:
+            await self.tree.sync(guild=discord.Object(self.testing_guild_id))
+
+        # This would also be a good place to connect to our database and
+        # load anything that should be in memory prior to handling events.
+
+
+async def main():
+
+    # When taking over how the bot process is run, you become responsible for a few additional things.
+
+    # 1. logging
+
+    # for this example, we're just going to use some basic defaults provided by python
+    # for more info on setting up logging, see https://docs.python.org/3/howto/logging.html
+    logging.basicConfig(level=logging.INFO)
```

Comment:
d.py has a dedicated logging page which is decently in-depth for those who just want quick examples, might want to refer to it too and use one of the snippets.

Reply:
Sure, this is a good place to advocate for devs to do more with logging than the basics anyway once they step into needing more fine-grained control over start.